### ISSUE TITLE
What's the parallelism story for ecs-rs?

### ISSUE THREAD

**Issue by olsonjeffery** [HeroesGrave/ecs-rs#10]
Looking at the codebase, I see that parallelism is a problem that is left to individual `System`s, without any library-level solution (hindered by the fact that all of the state-affecting parts of `ecs` are only ever exposed as `&mut`.

I'm curious if this is something that has been thought about, at all. With the current system constraints and API for state mutation, I image one solution would be marshall custom DTO instances of data that could be fed to another task that the `System` owns via a `TaskPool` it holds onto, or somesuch. This is a sort of no-shared-memory approach, but would be copy-intensive. Another drawback is all the boilerplate around the DTO objects that get sent across the task boundary and the individual `System` implementations having to own the roundtrip of the data (getting back the output from the off-task job and feeding it into the state-affecting mechanisms that are reachable from its task). Is there some other feasible approach?

One option would be to modify the API to expose cross-task-safe wrappers for `World` and/or its critical API calls (`modify|create`, etc), but that assumes that the internal data structures are themselves MT-safe (which they aren't).

Just wanted to start the conversation on this, if it hasn't already started. I really like what I see, so far, in this library and I am starting to use it in my own project.

**Comment by HeroesGrave** [comment]
Now that everything has been refactored, it may be possible to create a `SyncWorld` type that has the needed wrappers (Arc, Mutex) to have data shared safely over threads.

Regardless of implementation, multithreading will add several restrictions that I don't want to have on single-threaded instances, so it will have to be a separate implementation.

**Comment by viperscape** [comment]
concurrency in an ecs is complex and I think a singlethreaded approach is probably complex enough on its own, but I have given it some thought and [wrote down some ideas](https://github.com/viperscape/rust-scene/wiki/Management-of-Entities'-Components-as-seen-from-Systems), feel free to peruse

**Comment by mtsr** [comment]
I think that thread-per-system is too limited a scope in the long term. Especially since many systems have a lot of potential for internal parallelism in addition to parallelism between systems. Something like a task graph seems great here (either plain task scheduling or something like flow graph from Intel's threadingbuildingblocks.org).

Of course this also gets hugely complex. But if you want to keep things simple I would go for parallelism within systems only.

**Comment by viperscape** [comment]
I think you could probably still achieve the same parallelism with a thread devoted to a system if that system then used a task pool (either dedicated for the system, or the world); having the system on its own thread means it can also house all its own data which makes for partitioning up the work simple (eg: no mutex), pseudo code below:

```rust
let sys_clsr = |x| x+1;
let bd = &[ent_data]; //slice of some big entity data that the system owns
let vbd = partition(bd,5); //partition by 5 sections
let vr: Vec<ent_data>; //vec of results (maybe as future-channel-receivers or promises)
for n in vbd.drain() { vr.push(tp.send(n, sys_clsr)); } // taskpool receives work and fn to operate with, vec-results receives a future to await on
for n in vr.drain() { world.signal(vr.pop().unwrap().await()); } //wait on results, signal other systems if necessary
```

I like the dependency graph idea, reminds me of pistons behavior trees, as well a node-graph (which is what it looks like). In your [other post](https://github.com/HeroesGrave/ecs-rs/issues/13#issuecomment-77739944) you mention a query service, this sounds reasonable and I think is key to this all working. A query service could be adapted for both single and multithreaded I believe.

**Comment by mtsr** [comment]
Sure, but you would also get lots of contention if many systems to their own parallelism.

**Comment by HeroesGrave** [comment]
After lots of thought around the potential design of such functionality, I've decided that it won't happen in *this* library.

This library is focused on single-threaded performance, which means running through each individual system as fast as possible in sequence.

A parallel ecs would require processing as many systems as possible at the same time (well, up to the number of CPU cores available). To do this would require some tricky scheduling that would (presumably) ruin single-threaded performance.

However, a parallel ecs is still something I want to look into in the future. But it won't be as part of this library unless it turns out that it beats single-threaded performance anyway.

So for now, closed.

**Comment by minecrawler** [comment]
Wow, this crate really looks awesome! I was looking for a usable system just like this one!

Since I am working on a game, heavily relying on multithreading for various tasks, parallelism is very important for me. So when I saw this issue, I first was a bit sad, but then thought about it and came up with a solution I want to use in my game. This solution might be implemented into ecs-rs itself for convenience, so let me present it here:

- We only need one thread to distribute tasks. It does not make sense to split that up and other threads can start working on stuff while the first thread is still distributing.
- When using a work-queue, "distributing" tasks is as simple as filling up the queue with entities, data and closures to execute
- So what we can do is create a MT wrapper which creates a thread pool, then updates the world in the main thread and fills all entities, data and closures into a (linked) list.
- Threads in the pool implement a loop which checks the work-queue. If there is an item, the thread should consume the first item (as an atomic operation, possibly use a mutex) and execute the closure with the entity and closure.
- You can then return execution to the API user so that they can do other stuff on the main thread. You might also provide a method which lets the main thread choin the thread pool and help work on the task-queue. Either way, you also need to let the user define a closurte which is executed once the work-queue is drained.

That way, ecs-rs can stay single-threaded, but also offer multi-threading convenience. Using the "on-queue-drained" event, a user can let the system work on input and calculations, then render (using Vulkan or Direct3D 12, which are MT-ready) without a big hassle (using some kind of state which defines the current phase).
